Nimble table hld (#812)
ni/nimble: packages/nimble-components/src/table/specs/README.md

> # Nimble Table High Level Design

Reviewer: I envision a few more PRs covering design decisions like:
 - element API
 - performance testing
 - accessibility
 - everything else in our [custom component template](https://github.com/ni/nimble/blob/main/specs/templates/custom-component.md) (this might be split across several PRs)

I'm trying to envision the right document breakdown for those. 

1. Will all of it go in one mega spec, with this being the first set of content? If so then maybe we should format this document using the custom-component template.
2. Or will we have separate documents? If so then maybe we should call this `table-tech-stack.md`

I'm also OK to defer this decision until we make our next design PR, and reformat/rename this document then if needed.

## Problem Statement

Tables are used widely throughout NI, and are a common component in design systems. Third-party tables currently in use within NI web products have been unsatisfactory on multiple fronts, and so we have found it necessary to introduce a table into the `nimble-components` library.

## Links To Relevant Work Items and Reference Material

-   [Nimble Table Issue](https://github.com/ni/nimble/issues/283)
-   [Nimble Table Research Issue](https://github.com/ni/nimble/issues/285)
-   [Nimble Table Project Page](https://github.com/orgs/ni/projects/11)
-   [Nimble Table Research Lucid Board](https://lucid.app/lucidspark/31f2314d-dd8e-46fd-8fc1-6e9f66700bb3/edit?viewport_loc=-6023%2C-26311%2C35411%2C20921%2CloaYwcZLRray4&invitationId=inv_38839ad5-72b2-4975-ab7a-6d8be33c960c)

## Implementation / Design

We have decided to build our table using [TanStack Table](https://tanstack.com/table/v8) as the data model, and create the UI ourselves using typical FAST-based templates. Using TanStack Table provides a lot of lift for the state management/API of the table, including:

-   Row selection

> -   Column sizing/sorting/orderin/pinning

Reviewer: ```suggestion
-   Column sizing/sorting/ordering/pinning
```

## Alternative Implementations / Designs

### Perspective

[Perspective](https://perspective.finos.org/) is a framework for working with and visualizing table data in a highly performant way. By moving all of the various processing tasks, such as sorting and filtering, into a separate web-worker thread, it allows the UI to remain responsive.

Pros

-   Architecture allows visualization to work with a dataset in the millions of rows without performance impact.
-   Plugin system allowing for the creation of various views that can be swapped out to visualize the underlying data.
-   Well tested

Cons

-   Lack of support for hierarchical data

>     -   This results in the creation of a variety of state management in the table component itself, including grouping, sorting, and row selection. Some of which negates the performance benefits we would otherwise see for free.

Reviewer: nit: clarity 
```suggestion
    -   We could support hierarchical data by writing custom state management logic in the table component itself for features like grouping, sorting, and row selection. Some of which negates the performance benefits we would otherwise see for free.
```

-   Higher cost to implement a solution for, and more code to test/maintain.

> Ultimately we decided against using Perspective as one of its most appealing traits, its performance, would in many situations be neutralized by the implementation needed to support use cases like hierarchical data.

Reviewer 2: Since hierarchical data isn't extremely common across our tables, it might be good to give a different example (such as selection) that would be very common. If the only performance impact was on hierarchical data, I think there would have been a more compelling reason to still use perspective with different performance caps for hierarchical data than non-hierarchical data.

Author: Updated.

> Ultimately we decided against using Perspective as one of its most appealing traits, its performance, would in many situations be neutralized by the implementation needed to support use cases like row selection and hierarchical data.

Reviewer 3: I'd like to inject a thought in there but don't feel strongly about it

```suggestion
Ultimately we decided against using Perspective as the general purpose table as one of its most appealing traits, its performance, would in many situations be neutralized by the implementation needed to support use cases like row selection and hierarchical data.

We do see future potential for leveraging Perspective for more targeted data analysis use-cases like:

- as a parametric data analysis engine with multi-visualization cross-filtering
- client side visualization and analysis of large data sets (100k - 10M)
- interactive analysis of massive server-side data sets (10M - 1B+).

There likely will be ways to leverage many of the general purpose table features (API, column configuration, styling, etc) in a theoretical future Perspective table if clients request that level of performance capability.
```

### Tabulator

[Tabulator](https://tabulator.info/) is a full-fledged table web component, offering a wide range of features that are well aligned with our needs.

Pros

-   Extensive feature set
-   Good documentation/examples

Cons

-   Virtualization doesn't meet our requirements

> -   Not Typescript-based

Reviewer: ```suggestion
-   Not TypeScript-based
```
